contracts/cw20-base/src/contract.rs

```diff
@@ -0,0 +1,337 @@
+use crate::error::ContractError;
+use crate::responses::{BalanceResponse, Cw20Coin, Cw20ReceiveMsg, TokenInfoResponse};
+use crate::validation::{validate_accounts, validate_msg, verify_logo};
+use cosmwasm_schema::cw_serde;
+use cosmwasm_std::{
+    ensure, Addr, Binary, BlockInfo, Deps, DepsMut, Env, MessageInfo, Response, StdError,
+    StdResult, Storage, Uint128,
+};
+use cw2::set_contract_version;
+use cw20_allowances::responses::AllowanceResponse;
+use cw20_marketing::responses::{LogoInfo, MarketingInfoResponse};
+use cw20_marketing::Logo;
+use cw20_minting::responses::MinterResponse;
+use cw_storage_plus::{Item, Map};
+use sylvia::{contract, schemars};
+
+// version info for migration info
+const CONTRACT_NAME: &str = "crates.io:cw20-base";
+const CONTRACT_VERSION: &str = env!("CARGO_PKG_VERSION");
+
+#[cw_serde]
+pub struct TokenInfo {
+    pub name: String,
+    pub symbol: String,
+    pub decimals: u8,
+    pub total_supply: Uint128,
+    pub mint: Option<MinterData>,
+}
+
+#[cw_serde]
+pub struct MinterData {
+    pub minter: Addr,
+    /// cap is how many more tokens can be issued by the minter
+    pub cap: Option<Uint128>,
+}
+
+#[cw_serde]
+pub struct InstantiateMarketingInfo {
+    pub project: Option<String>,
+    pub description: Option<String>,
+    pub marketing: Option<String>,
+    pub logo: Option<Logo>,
+}
+
+#[cw_serde]
+pub struct InstantiateMsgData {
+    pub name: String,
+    pub symbol: String,
+    pub decimals: u8,
+    pub initial_balances: Vec<Cw20Coin>,
+    pub mint: Option<MinterResponse>,
+    pub marketing: Option<InstantiateMarketingInfo>,
+}
+
+pub struct Cw20Base<'a> {
+    pub(crate) token_info: Item<'static, TokenInfo>,
+    pub(crate) marketing_info: Item<'static, MarketingInfoResponse>,
+    pub(crate) logo: Item<'static, Logo>,
+    pub(crate) balances: Map<'static, &'a Addr, Uint128>,
+    pub(crate) allowances: Map<'static, (&'a Addr, &'a Addr), AllowanceResponse>,
+    // TODO: After https://github.com/CosmWasm/cw-plus/issues/670 is implemented, replace this with a `MultiIndex` over `ALLOWANCES`
+    pub(crate) allowances_spender: Map<'static, (&'a Addr, &'a Addr), AllowanceResponse>,
```

Comment:
There is a good question if there should be `pub` or `pub(crate)` to make it most useful. There are good arguments for both. For now, leave it as it is, but we need to figure out the proper style in future having it in mind.